# Implementation of T-test, Anova, Mann-whitney-U, Kruskal-Walls and Cohen's D

# T-Test

Use when you have two independent groups and the numerical variables

In [13]:
from scipy.stats import ttest_ind

In [14]:
group1 = [10, 12, 14, 15, 13]
group2 = [18, 20, 17, 19, 21]

In [15]:
t_stat, p_value = ttest_ind(group1, group2)

print(f"T-statistics: {t_stat}")
print(f"P-Value: {p_value}")

T-statistics: -5.567764362830021
P-Value: 0.0005298298648541324


- P-Value: 0.0005298298648541324. It rejects the null hypothesis; move to the alternative. Hence, there is a significant relationship between the two variables

- T-statistics: It tells you how far the result is from H0. The negative sign means that the observed difference in the negative directions relative to null hypothesis



- Use : 2 groups + numerical data + approximately normal data.

# ANOVA Test (Analysis of Variance)

Use when you have three or more independent groups and a numerical variable.


In [5]:
from scipy.stats import f_oneway

group1 = [10, 12, 14, 15, 13]
group2 = [18, 20, 17, 19, 21]
group3 = [25, 23, 27, 26, 24]

f_stat, p_value = f_oneway(group1, group2, group3)

print(f"F-statistics: {f_stat}")
print(f"P-value: {p_value}")

F-statistics: 64.16091954022987
P-value: 3.9114327553428166e-07


- The ANOVA result provides strong statistical evidence against the hypothesis that all group means are equal.

- F-statistic: 64.16091954022987, means the variation between the group means is much larger than the variation within the groups.

# Mann–Whitney U test

- This is a non-parametric alternative to the independent t-test.

- Use it for two independent groups when the numerical/ordinal data do not meet the assumptions of a t-test.


In [9]:
from scipy.stats import mannwhitneyu

group1 = [10, 12, 14, 15, 13]
group2 = [18, 20, 17, 19, 21]

u_stat, p_value = mannwhitneyu(
    group1,
    group2,
    alternative='two-sided'
)

print("U-statistic:", u_stat)
print("p-value:", p_value)

U-statistic: 0.0
p-value: 0.007936507936507936


Use: 2 independent groups + non-normal/ordinal data.

# Kruskal–Wallis test

- This is a nonparametric alternative to one-way ANOVA.

- Use it for 3 or more independent groups when ANOVA assumptions aren't met.

In [11]:
from scipy.stats import kruskal

group1 = [10, 12, 14, 15, 13]
group2 = [18, 20, 17, 19, 21]
group3 = [25, 23, 27, 26, 24]

h_stat, p_value = kruskal(
    group1,
    group2,
    group3
)

print("H-statistic:", h_stat)
print("p-value:", p_value)

H-statistic: 12.5
p-value: 0.0019304541362277095


- Use: 3+ independent groups + non-normal/ordinal data.

# Cohen's D

- It is an effect size used to describe the standardised difference between two group means.

- How large is the difference?

For two independent groups:

$$
d = \frac{\bar{X}_1 - \bar{X}_2}{s_p}
$$


In [1]:
import numpy as np

group1 = np.array([10, 12, 14, 15, 13])
group2 = np.array([18, 20, 17, 19, 21])

mean1 = np.mean(group1)
mean2 = np.mean(group2)

n1 = len(group1)
n2 = len(group2)

sd1 = np.std(group1, ddof=1)
sd2 = np.std(group2, ddof=1)

# Pooled standard deviation
sp = np.sqrt(
    ((n1 - 1) * sd1**2 + (n2 - 1) * sd2**2)
    / (n1 + n2 - 2)
)

# Cohen's d
d = (mean1 - mean2) / sp

print("Cohen's d:", d)

Cohen's d: -3.521363372331801


Cohen's 
d
=
−
3.52
, indicating a very large standardized difference between the two groups, with Group 1 having a lower mean than Group 2.

